examples/nn.dx

```diff
@@ -0,0 +1,242 @@
+' # Neural Networks
+
+include "plot.dx"
+
+' ## NN Prelude
+
+
+def relu (input : Float) : Float =
+  select (input > 0.0) input 0.0
+
+instance [Add a, Add b] Add (a & b)
+  add = \(a, b) (c, d). ( (a + c), (b + d))
+  sub = \(a, b) (c, d). ( (a - c), (b - d))
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward: param -> inp -> out &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
```

Comment:
I think `Layer` could be defined as an `interface`, which would save you having to write these getter functions.

Then things like `Dense` would just be instances of `Layer`, and more things would be named where they are defined.   I haven't thought this through though, there might be some downside I haven't thought of.

Reply:
`Layer` as an interface makes complete sense to me. Especially if/when associated types are supported.

Here's `Layer` from Swift for Deep Learning:
```swift
// The layer interface.
// It's a differentiable function.
protocol Layer: Differentiable {
  associatedtype Input: Differentiable
  associatedtype Output: Differentiable
  @differentiable(wrt: input)
  @differentiable(wrt: (self, input))
  func callAsFunction(_ input: Input) -> Output
}

// Dense: a `Layer` conforming type.
// dense(x) = wx + b
struct Dense: Layer {
  var weight, bias: Float
  func callAsFunction(_ input: Float) -> Float { weight * input + bias }
}

// Example usage.
let dense = Dense(weight: 2021, bias: 14)
print(dense(Float.pi)) // 6363.1587
```

Reply:
I tried so hard to do this, but I couldn't figure it out. 

I don't think `interface` in dex allows you to look at the types, and if I try to wrap the parameters in a `data` it breaks `vjp`.

Reply:
Just to make it more explicit. You want something like -> 

```haskell
interface Layer a  
    forward : (Param a) -> (Input a) ->  (Output a)
    init : Key -> (Param a)
```

This seems like something you could do in haskell (if you turned on lots of flags), but is beyond Dex.

Reply:
I should have known you had already tried to do this!  Hmmm, I think there still might be a way forward by making use of dependent types, but I'm getting confused by the different layers of abstraction.

Reply:
In this example, what would `a` be? Is it something like `data Dense a b`? I'm wondering in that case what data Dense would actually carry around.

In other words, maybe there's a sense in which a specific type of layer isn't itself a datatype, it's just a bundle of functions? In which case it does kinda make sense for `Layer` to be the datatype, and `dense` to be a value of type `Layer`? As opposed to `Dense` being a type, which implies that there are multiple values of `Dense`.

(relatedly, right now, `Dense` is an alias for a type of `Layer`. But there could be layer implementations that conform to the same API as `Dense` but aren't "values of Dense", right? So maybe `Dense` shouldn't even be a type?

Reply:
The other thing you might imagine doing would be for `Dense` to be a datatype carrying around _the parameters_? But knowing how to init them and how to apply them to data? So you'd have something like
```
interface Layer a inp out
    forward : a -> inp -> out
    init : Key -> a
```
Although I'm not sure if maybe we'd want functional-dependency-style inferring of `inp` and `out`, maybe like
```
interface Layer a
    imp : Type
    out : Type
    forward : a -> inp -> out
    init : Key -> a
```
(this isn't currently supported, although I have some ideas)

Reply:
My original thought was exactly this. Dense should hold the parameters, init the parameters, and even hide the parameters in its type. However this idea ran into several hurdles 

* `grad` only works on raw values / tuples. 
* You need to be able to unpack / repack data structures of params (I couldn't figure out how to do that)
* Interfaces don't allow multiple types as you have noted 

Would love some help if you have ideas about how to do this. 

That being said, I don't hate the current API (Dense being an alias). It roughly does what I was aiming for => reuse the verbs `forward` and `init`, make it easy to override init, cleanly type the input and output of layers, avoid rewriting any types more than once, and have an intermediate function if case you need to wire through Key's or other state. 

The only thing I don't love is that the type of the outer model is super nasty, and exposes all the params.

Reply:
1 sounds like something we want to eventually support (especially if the type is just a newtype wrapper around something we can take the grad of).

Could you expand more on 2? I don't quite see what the problem is with unpacking/repacking.

My only real complaint with the API as you've defined it is that I don't see why `Dense` needs to be an alias at all. It seems like morally it's better to say `dense (a:Type) (b:Type) : Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float)`, since the type of "thing" that `dense` returns isn't "one of many values of Dense" it's "one of many values of Layer". There's essentially only one dense object, right?

Reply:
Regarding "the type of the outer model is super nasty", @dougalm and I at one point discussed a high-level idea for reversible type aliases, where the pretty printer could try to show the aliases that the user used (in a best-effort basis) instead of showing the fully-expanded version. Maybe that would help here?

Reply:
1) I agree. But morally is it bad to have an alias? It makes the type clearer, and it is helpful in debugging since you can specify the type with less keystroke. (For instance below in the lenet you could type the modules.) 

2) My original design had a datastructure

```haskell
data Param a =
        PreInit Key => a | ActiveParam a

fromActive (a:Param) = case a of ActiveParam a' -> a'  
```

And I was imaging that you would have a bunch of tuples of these things in each Layer. 

But then to run grad and avoid this data structure, you need to convert an arbitrary tuple of Params to a tuple of tables, then thread it through a function, updat. But I couldn't figure out how to do this in the current type system.

Reply:
Yeah, I think showing the aliases would be neat? Would be cool for naming `Height = Fin 8` as well. 

I'm not sure why I even care that the type has all the params in it. It's not a practical problem. The biggest issue is that the type errors get extremely long if you use aliases.

Reply:
Ah, that's interesting about `Param`. You almost want a sort of `jax.tree_map` for structures of params or something? If Dex had more powerful typeclasses, maybe you could write a typeclass for that kind of traversal. But agreed that it seems difficult (maybe impossible) to do now.

Reply:
Yeah that was what I was thinking. But probably because that's what I'm used to. In a Dex world recursive structures might just be less important.

Reply:
Well they are important, just not implemented, because dealing with recursion is hard! I think that @danieldjohnson has correctly identified a few important features that we're lacking, but would greatly help here:
1. Associated types in interfaces
2. Preventing alias expansion when pretty printing types
I'll open issues for those!